libkbfs/unflushed_path_cache.go

```
@@ -0,0 +1,328 @@
+					"already processing one", irmd.Revision())
+			}
+
+			log.CDebugf(ctx, "Skipping unflushed paths for revision %d "+
+				"since it's already in the cache", irmd.Revision())
+			continue
+		}
+
+		processedOne = true
+		winfo := writerInfo{
+			uid:      uid,
+			kid:      kid,
+			revision: irmd.Revision(),
+			// There won't be any conflicts, so no need for the
+			// username/devicename.
+		}
+		err := chains.addOps(codec, irmd.data, winfo, irmd.localTimestamp)
+		if err != nil {
+			return err
+		}
+	}
+	if !processedOne {
+		return nil
+	}
+
+	chains.mostRecentMD = irmds[len(irmds)-1]
+
+	err := cpp.populateChainPaths(ctx, log, chains, true)
+	if err != nil {
+		return err
+	}
+
+	for _, chain := range chains.byOriginal {
+		for _, op := range chain.ops {
+			revPaths := unflushedPaths[op.getWriterInfo().revision]
+			if revPaths == nil {
+				revPaths = make(map[string]bool)
+				unflushedPaths[op.getWriterInfo().revision] = revPaths
+			}
+			revPaths[op.getFinalPath().String()] = true
+		}
+	}
+	return nil
+}
+
+// prepUnflushedPaths returns a set of paths that were updated in the
+// given revision.
+func (upc *unflushedPathCache) prepUnflushedPaths(ctx context.Context,
+	uid keybase1.UID, kid keybase1.KID, codec kbfscodec.Codec,
+	log logger.Logger, irmd ImmutableRootMetadata) (
+	unflushedPathsPerRevMap, error) {
+	cpp := func() chainsPathPopulator {
+		upc.lock.Lock()
+		defer upc.lock.Unlock()
+		return upc.chainsPopulator
+	}()
+
+	// The unflushed paths haven't been initialized yet.
+	if cpp == nil {
+		return nil, nil
+	}
+
+	newUnflushedPaths := make(unflushedPathsMap)
+	irmds := []ImmutableRootMetadata{irmd}
+
+	err := addUnflushedPaths(ctx, uid, kid, codec, log, irmds, cpp,
+		newUnflushedPaths)
+	if err != nil {
+		return nil, err
+	}
+	if len(newUnflushedPaths) > 1 {
+		return nil, fmt.Errorf("%d unflushed revisions on a single put",
+			len(newUnflushedPaths))
+	}
+
+	return newUnflushedPaths[irmd.Revision()], nil
```

if `addUnflushedPaths` somehow fills in the wrong key, this will silently return nil. Perhaps defensively check for that case and return an error/panic?